### PR TITLE
Stabilize JavaScript Profiler in Cobalt

### PR DESCRIPTION
**Pull request by aelzeiny**
b/323983545

A major refactor was needed in order to prevent the JavaScript Profiler from SEGFAULTING due to a garbage-collected nullptr.

This refactor includes the use of a `ProfilerGroup` class. The life of a cobalt::js_profiler::Profiler is extended by adding scoped_refptrs to the ProfilerGroup; thus preventing GC. Profilers live as long as their ProfilerGroup, which lives as long as the Isolate and Web Agent. One `ProfilerGroup` is bound to one `web::Agent`, and contains one `v8::Isolate`, one `v8::CpuProfiler`, and many `cobalt::js_profiler::Profiler`s. 

The major advantage of this refactor is that it does not crash, and is more faithful to the [original Chromium Implementation](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/timing/profiler_group.cc;bpv=0;bpt=0).

Test-On-Device: true

### CODE REVIEW

File: cobalt/js_profiler/profiler.cc
Comment (kaidokert) [inline cobalt/js_profiler/profiler.cc:15]: Tiny nit: commit noise. ( unless pre-commit added it )

File: cobalt/js_profiler/profiler.cc
Comment (kaidokert) [inline cobalt/js_profiler/profiler.cc:28]: This is in Cobalt namespace, don't use Starboard logging directly. Use the LOG macros that come from //base logging. So SB_LOG below should just be LOG

File: cobalt/js_profiler/profiler.h
Comment (kaidokert) [inline cobalt/js_profiler/profiler.h:15]: nit: commit noise

File: cobalt/js_profiler/profiler.idl
Comment (aelzeiny) [reply cobalt/js_profiler/profiler.idl:23]: Filed a bug: b/326337485

**Comment by aelzeiny** [comment]
Merge [PR for black box tests](https://github.com/youtube/cobalt/pull/2468) after merging this.

File: cobalt/js_profiler/profiler_group.cc
Comment (jellefoks) [inline cobalt/js_profiler/profiler_group.cc:1]: 2024

File: cobalt/js_profiler/profiler.cc
Comment (aelzeiny) [reply cobalt/js_profiler/profiler.cc:81]: Added notes to b/326337485

Comment (aelzeiny) [reply cobalt/js_profiler/profiler.cc:81]: For what it's worth, I tried this. There's no easy way to get to the ScriptValue in EventTarget or EventTargetListenerInfo without some meaningful refactors. I'll document the shortcomings of this hacky approach and revisit later.

File: cobalt/js_profiler/profiler_group.h
Comment (aelzeiny) [reply cobalt/js_profiler/profiler_group.h:1]: Done.

File: cobalt/js_profiler/profiler.cc
Comment (aelzeiny) [reply cobalt/js_profiler/profiler.cc:84]: Fixed to allow for multiple callbacks. Also added documentation to the IDL files.

File: cobalt/js_profiler/profiler.cc
Comment (aelzeiny) [reply cobalt/js_profiler/profiler.cc:119]: Ack.

File: cobalt/js_profiler/profiler.idl
Comment (aelzeiny) [reply cobalt/js_profiler/profiler.idl:23]: This approach is by all standards "hacky" but necessary. I've filed a bug (b/326337485) and added a disclaimer to profiler.idl denoting the clear disadvantages of the given approach.

File: cobalt/js_profiler/profiler.idl
Comment (aelzeiny) [reply cobalt/js_profiler/profiler.idl:27]: I noticed and acknowledge the hacky approach. b/326337485 hopefully explains why this approach was chosen. As requested, I've updated `profiler.idl` to caution users who are implementing this API.

**Review by jellefoks — APPROVED** [review]
Ok, that should at least reduce the amount of surprise by future users.

**Comment by aelzeiny** [comment]
Test-On-Device: true
b/323983545

**Review by kaidokert — APPROVED** [review]
lgtm, provided integration tests in cl/608835107 and https://github.com/youtube/cobalt/pull/2468 also land. But no need to hold up submitting this.

**Pull request merged**
